### ISSUE TITLE
[pickers] Infinite render loop when using `dayjs.tz`

### ISSUE THREAD

**Issue by Samaa-Amin** [mui/mui-x#5876]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

Date pickers throughs error on certain dates with certain time zones ex: ("2010-04-30 12:00", "Egypt"):
`Too many re-renders. React limits the number of renders to prevent an infinite loop.` 

### Expected behavior 🤔

Should be able to pick any date with any time zone without errors

### Steps to reproduce 🕹

Steps:
https://codesandbox.io/s/cool-hooks-zpmyro?file=/src/App.tsx:512-541

1. pick this date "2010-04-30" from the date time picker field

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by flaviendelangle** [comment]
It does not occur with the other days I tested
And I can reproduce it in a very simple test case

[Simplified reproduction with timezone => crashes](https://codesandbox.io/s/quirky-sound-9r3jr9?file=/src/App.tsx)

I does not crash if I replace `dayjs.tz` by `dayjs` so it seems to be related to the timezone management in dayjs that we probably don't support correctly.
I'll have a deeper look

_____

By the way, you should not have to re-parse your date in the `dayjs` format in your `onChange` handler.

**Comment by flaviendelangle** [comment]
I think the bug in in dayjs https://github.com/iamkun/dayjs/issues/1736

The infinite loop is caused by the following `if` being `true` even when the date stored in state and the date passed as props are the same

https://github.com/mui/mui-x/blob/33cf86590cd69cdcd94dbaa831c1aeb2e70785c0/packages/x-date-pickers/src/internals/hooks/usePickerState.ts#L258:L260

For `DateTimePicker`, `valueManager.areValuesEqual` just calls the `isEqual` method of the adapter

https://github.com/mui/mui-x/blob/33cf86590cd69cdcd94dbaa831c1aeb2e70785c0/packages/x-date-pickers/src/DateTimePicker/shared.ts#L130

And the `dayjs` adapter calls the `isSame` of the `dayjs` object.

https://github.com/dmtrKovalenko/date-io/blob/b39210ed1114bc2d03bb487b94e1e32b1c7a8a8b/packages/dayjs/src/dayjs-utils.ts#L345:L351

The problem being that this `isSame` method returns false even when comparing some date with itself.

I had a look at the `isSame` method and the problem seems to be in the second half of the comparison

https://github.com/iamkun/dayjs/blob/6c2a51702333c7846c90036bd5a6e3347bb2939e/src/index.js#L115:L118

With the 30th of April (and any other summer time date I think), `a <= this.endOf(a) === false` 

I think the problem is that `this.endOf(a)` changes the offset of the date.

____

TLDR: Unless I am missing something, the problem is unfortunately on our side
There are open issues of the `dayjs` repository to upvote but not sure we can do better than that.

**Comment by msalahz** [comment]
I'm having the same issue. Are there any suggestions or workarounds?

**Comment by flaviendelangle** [comment]
The best solution would clearly be to help the `dayjs` team fix the bug
I don't see any workarounds but I'm might be missing some.

**Comment by flaviendelangle** [comment]
@LukasTy if you have some idea

**Comment by fungaldore** [comment]
I'm working around this by converting to a javascript date object to be used in the picker, then back again for the rest of my app.  I have the MUI Date Picker in its own component, and looks similar to:
```
const MyDatePicker = ({ value, handleChange, timezone }) => {
  const dayjsToDatefns = (dayjsDate) => dayjsDate.toDate();
  const datefnsToDayjs = (datefnsDate) => dayjs(datefnsDate).tz(timezone);

  return (
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <MobileDatePicker
        onChange={(newVal) => handleChange(datefnsToDayjs(newVal))}
        renderInput={(props) => <TextField {...props} />}
        value={dayjsToDatefns(value)}
      />
    </LocalizationProvider>
  );
};
```
Now I can use DayJS in the rest of my app like:
`const [startDate, setStartDate] = useState(dayjs().tz('America/Denver', true));`
and
`<MyDatePicker value={startDate} handleChange={setStartDate} timezone={'America/Denver'} />`

**Comment by Thebarda** [comment]
Hi, I don't what version of mui-x, you are using, I've fixed the error by removing the `.tz(timezone)` from the value prop. But I had to override the Adapter in order to manage timezones and DSTs. Ping if you are interested by Adapter override

**Comment by flaviendelangle** [comment]
With #8261, we not longer use `isSame` in the `isEqual` method, so the component no longer crashes :+1:
